File: apps/agent/app/tools/canvas.py
from __future__ import annotations

"""
Canvas Document Tools

These tools manage sidebar canvas documents - a scratchpad where users can type notes
and AI can read/modify content. Documents are stored locally and synced via IPC to desktop.
"""

import uuid
from datetime import datetime
from typing import Any, Dict, Optional


# In-memory storage for canvas documents
# In production, these are forwarded to the desktop app via IPC
_DOCUMENTS: Dict[str, Dict[str, Any]] = {}


def _gen_id() -> str:
    return f"doc-{uuid.uuid4().hex[:8]}"


def _now_iso() -> str:
    return datetime.utcnow().isoformat() + "Z"


async def canvas_list(args: Dict[str, Any], emit: Optional[Any] = None) -> Dict[str, Any]:
    """List all canvas documents."""
    documents = list(_DOCUMENTS.values())
    return {"ok": True, "documents": documents}


async def canvas_read(args: Dict[str, Any], emit: Optional[Any] = None) -> Dict[str, Any]:
    """Read content from a canvas document."""
    doc_id = args.get("documentId")
    
    if doc_id:
        doc = _DOCUMENTS.get(doc_id)
        if not doc:
            return {"ok": False, "error": "Document not found", "document": None}
        return {"ok": True, "document": doc}
    
    # If no ID provided, return the most recently updated document
    if not _DOCUMENTS:
        return {"ok": True, "document": None}
    
    most_recent = max(_DOCUMENTS.values(), key=lambda d: d.get("updatedAt", ""))
    return {"ok": True, "document": most_recent}


async def canvas_write(args: Dict[str, Any], emit: Optional[Any] = None) -> Dict[str, Any]:
    """Write or modify content in a canvas document."""
    doc_id = args.get("documentId")
    content = args.get("content")
    title = args.get("title")
    action = args.get("action", "replace")
    position = args.get("position", 0)
    
    # Find target document
    if doc_id:
        doc = _DOCUMENTS.get(doc_id)
        if not doc:
            return {"ok": False, "error": "Document not found"}
    else:
        # Use most recent document or create new one
        if not _DOCUMENTS:
            # Auto-create a document
            doc_id = _gen_id()
            now = _now_iso()
            doc = {
                "id": doc_id,
                "title": title or "Untitled",
                "content": "",
                "createdAt": now,
                "updatedAt": now,
            }
            _DOCUMENTS[doc_id] = doc
        else:
            doc = max(_DOCUMENTS.values(), key=lambda d: d.get("updatedAt", ""))
    
    # Apply content changes
    if content is not None:
        current_content = doc.get("content", "")
        if action == "replace":
            doc["content"] = content
        elif action == "append":
            doc["content"] = current_content + content
        elif action == "insert":
            pos = max(0, min(position, len(current_content)))
            doc["content"] = current_content[:pos] + content + current_content[pos:]
    
    # Handle edit action (find and replace)
    if action == "edit":
        old_str = args.get("old_string")
        new_str = args.get("new_string", "")
        current_content = doc.get("content", "")
        
        if not old_str:
            return {"ok": False, "error": "old_string is required for edit action"}
        
        if old_str not in current_content:
            return {"ok": False, "error": "old_string not found in document"}
        
        doc["content"] = current_content.replace(old_str, new_str)
    
    # Update title if provided
    if title is not None:
        doc["title"] = title
    
    doc["updatedAt"] = _now_iso()
    
    if emit:
        try:
            await emit("progress", {"kind": "canvas_write", "documentId": doc["id"], "action": action})
        except Exception:
            pass
    
    return {"ok": True}


async def canvas_create(args: Dict[str, Any], emit: Optional[Any] = None) -> Dict[str, Any]:
    """Create a new canvas document."""
    title = args.get("title", "Untitled")
    content = args.get("content", "")
    
    doc_id = _gen_id()
    now = _now_iso()
    
    doc = {
        "id": doc_id,
        "title": title,
        "content": content,
        "createdAt": now,
        "updatedAt": now,
    }
    _DOCUMENTS[doc_id] = doc
    
    if emit:
        try:
            await emit("progress", {"kind": "canvas_create", "documentId": doc_id})
        except Exception:
            pass
    
    return {"ok": True, "documentId": doc_id}


async def canvas_delete(args: Dict[str, Any], emit: Optional[Any] = None) -> Dict[str, Any]:
    """Delete a canvas document by ID."""
    doc_id = args.get("documentId")
    
    if not doc_id:
        return {"ok": False, "error": "documentId is required"}
    
    if doc_id not in _DOCUMENTS:
        return {"ok": False, "error": "Document not found"}
    
    del _DOCUMENTS[doc_id]
    
    if emit:
        try:
            await emit("progress", {"kind": "canvas_delete", "documentId": doc_id})
        except Exception:
            pass
    
    return {"ok": True}

File: apps/desktop/src/main/tools/handlers/canvas.ts
/**
 * Canvas Document Tool Handlers
 * 
 * These handlers manage sidebar canvas documents using the same
 * storage as the IPC handlers (canvas-documents.json).
 */

import { app, BrowserWindow } from 'electron';
import * as fs from 'fs';
import * as path from 'path';
import { RouterContext } from '../types';

// Canvas document storage path
const canvasDocsPath = (): string => {
  const userDataPath = app.getPath('userData');
  return path.join(userDataPath, 'canvas-documents.json');
};

// Load documents from disk
const loadCanvasDocs = (): any[] => {
  try {
    const p = canvasDocsPath();
    if (fs.existsSync(p)) {
      return JSON.parse(fs.readFileSync(p, 'utf-8'));
    }
  } catch (e) {
    console.warn('[canvas] Failed to load canvas documents:', e);
  }
  return [];
};

// Save documents to disk
const saveCanvasDocs = (docs: any[]): void => {
  try {
    fs.writeFileSync(canvasDocsPath(), JSON.stringify(docs, null, 2), 'utf-8');
  } catch (e) {
    console.warn('[canvas] Failed to save canvas documents:', e);
  }
};

// Get sidebar window for sending updates
const getSidebarWindow = (): BrowserWindow | null => {
  try {
    const { getSidebarWindow: getSidebar } = require('../../windows');
    return getSidebar();
  } catch {
    return null;
  }
};

/**
 * List all canvas documents
 */
export async function execCanvasList(_args: any, _ctx: RouterContext): Promise<any> {
  const documents = loadCanvasDocs();
  return { ok: true, documents };
}

/**
 * Read a canvas document by ID, or the most recent if no ID provided
 */
export async function execCanvasRead(args: any, _ctx: RouterContext): Promise<any> {
  const docId = args?.documentId;
  const docs = loadCanvasDocs();

  if (docId) {
    const doc = docs.find((d: any) => d.id === docId);
    if (!doc) {
      return { ok: false, error: 'Document not found', document: null };
    }
    return { ok: true, document: doc };
  }

  // Return most recent document if no ID specified
  if (docs.length === 0) {
    return { ok: true, document: null };
  }

  // Sort by updatedAt descending and return first
  const sorted = [...docs].sort((a, b) => {
    const aTime = new Date(a.updatedAt || a.createdAt || 0).getTime();
    const bTime = new Date(b.updatedAt || b.createdAt || 0).getTime();
    return bTime - aTime;
  });

  return { ok: true, document: sorted[0] };
}

/**
 * Write to a canvas document (replace, append, or insert content)
 */
export async function execCanvasWrite(args: any, _ctx: RouterContext): Promise<any> {
  const docId = args?.documentId;
  const content = args?.content;
  const title = args?.title;
  const action = args?.action || 'replace';
  const position = args?.position || 0;

  const docs = loadCanvasDocs();
  let doc: any;
  let docIndex: number;

  if (docId) {
    docIndex = docs.findIndex((d: any) => d.id === docId);
    if (docIndex < 0) {
      return { ok: false, error: 'Document not found' };
    }
    doc = docs[docIndex];
  } else {
    // Use most recent document or create new one
    if (docs.length === 0) {
      // Auto-create a document
      const newId = `canvas_${Date.now()}`;
      const now = new Date().toISOString();
      doc = {
        id: newId,
        title: title || 'Untitled',
        content: '',
        createdAt: now,
        updatedAt: now,
      };
      docs.unshift(doc);
      docIndex = 0;
    } else {
      // Sort by updatedAt and use most recent
      const sorted = docs.map((d, i) => ({ doc: d, idx: i })).sort((a, b) => {
        const aTime = new Date(a.doc.updatedAt || a.doc.createdAt || 0).getTime();
        const bTime = new Date(b.doc.updatedAt || b.doc.createdAt || 0).getTime();
        return bTime - aTime;
      });
      docIndex = sorted[0].idx;
      doc = docs[docIndex];
    }
  }

  // Apply content changes
  if (content !== undefined && content !== null) {
    const currentContent = doc.content || '';
    if (action === 'replace') {
      doc.content = content;
    } else if (action === 'append') {
      doc.content = currentContent + content;
    } else if (action === 'insert') {
      const pos = Math.max(0, Math.min(position, currentContent.length));
      doc.content = currentContent.slice(0, pos) + content + currentContent.slice(pos);
    }
  }

  // Update title if provided
  if (title !== undefined && title !== null) {
    doc.title = title;
  }

  doc.updatedAt = new Date().toISOString();
  docs[docIndex] = doc;
  saveCanvasDocs(docs);

  // Notify sidebar to update UI
  const sidebar = getSidebarWindow();
  if (sidebar && !sidebar.isDestroyed()) {
    sidebar.webContents.send('canvas:update', {
      documentId: doc.id,
      content: doc.content,
      title: doc.title,
      action,
      position,
    });
  }

  return { ok: true };
}

/**
 * Create a new canvas document
 */
export async function execCanvasCreate(args: any, _ctx: RouterContext): Promise<any> {
  const title = args?.title || 'Untitled';
  const content = args?.content || '';

  const newId = `canvas_${Date.now()}`;
  const now = new Date().toISOString();

  const doc = {
    id: newId,
    title,
    content,
    createdAt: now,
    updatedAt: now,
  };

  const docs = loadCanvasDocs();
  docs.unshift(doc);
  saveCanvasDocs(docs);

  return { ok: true, documentId: newId };
}

/**
 * Delete a canvas document by ID
 */
export async function execCanvasDelete(args: any, _ctx: RouterContext): Promise<any> {
  const docId = args?.documentId;

  if (!docId) {
    return { ok: false, error: 'documentId is required' };
  }

  const docs = loadCanvasDocs();
  const idx = docs.findIndex((d: any) => d.id === docId);

  if (idx < 0) {
    return { ok: false, error: 'Document not found' };
  }

  docs.splice(idx, 1);
  saveCanvasDocs(docs);

  return { ok: true };
}
